container/raw/inotify_watcher.go

```diff
@@ -55,46 +51,51 @@ func (iw *InotifyWatcher) AddWatch(containerName, dir string) (bool, error) {
 	iw.lock.Lock()
 	defer iw.lock.Unlock()
 
-	alreadyWatched := iw.containersWatched[containerName]
+	cgroupsWatched, alreadyWatched := iw.containersWatched[containerName]
 
 	// Register an inotify notification.
-	if !iw.cgroupsWatched[dir] {
+	if !cgroupsWatched[dir] {
 		err := iw.watcher.AddWatch(dir, inotify.IN_CREATE|inotify.IN_DELETE|inotify.IN_MOVE)
 		if err != nil {
 			return alreadyWatched, err
 		}
-		iw.cgroupsWatched[dir] = true
+
+		if cgroupsWatched == nil {
+			cgroupsWatched = make(map[string]bool)
+		}
+		cgroupsWatched[dir] = true
 	}
 
 	// Record our watching of the container.
 	if !alreadyWatched {
-		iw.containersWatched[containerName] = true
+		iw.containersWatched[containerName] = cgroupsWatched
 	}
 	return alreadyWatched, nil
 }
 
-// Remove watch from the specified directory. Returns if the container was already being watched.
+// Remove watch from the specified directory. Returns if this was the last watch on the specified container.
 func (iw *InotifyWatcher) RemoveWatch(containerName, dir string) (bool, error) {
 	iw.lock.Lock()
 	defer iw.lock.Unlock()
 
-	alreadyWatched := iw.containersWatched[containerName]
+	cgroupsWatched := iw.containersWatched[containerName]
 
```

Comment:
If it returns nil, all the operations we use here are handled (get, delete, and len). Went ahead and made it explicit though to make it clearer.